src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -351,517 +433,1099 @@ namespace hiop
+                                       d_Ui_csr,
+                                       d_Ux_csr,
+                                       d_P,
+                                       d_Q,
+                                       handle_rf_);
+    cudaDeviceSynchronize();
+    sp_status_ = cusolverRfAnalyze(handle_rf_);
+    return 0;
   }
 
   int hiopLinSolverNonSymSparseCUSOLVER::matrixChanged()
   {
     assert(n_ == M_->n() && M_->n() == M_->m());
-    assert(n_>0);
+    assert(n_ > 0);
 
     nlp_->runStats.linsolv.tmFactTime.start();
 
     if(!kRowPtr_) {
       this->firstCall();
     } else {
       // update matrix
-      for(int k=0;k<nnz_;k++) {
+      for(int k = 0; k < nnz_; k++) {
         kVal_[k] = M_->M()[index_covert_CSR2Triplet_[k]];
       }
-      for(int i=0;i<n_;i++) {
+      for(int i = 0; i < n_; i++) {
         if(index_covert_extra_Diag2CSR_[i] != -1) {
-          kVal_[index_covert_extra_Diag2CSR_[i]] += M_->M()[M_->numberOfNonzeros() - n_ + i];
-        }     
+          kVal_[index_covert_extra_Diag2CSR_[i]]
+              += M_->M()[M_->numberOfNonzeros() - n_ + i];
+        }
+      }
+    } // else
+    if((Numeric_ == nullptr) && (factorizationSetupSucc_ == 0)) {
+      Numeric_ = klu_factor(kRowPtr_, jCol_, kVal_, Symbolic_, &Common_);
+      if(Numeric_ == nullptr) {
+        nlp_->log->printf(hovSummary, "Numeric factorization failed. Regularizing ...\n");
+        // This is not a catastrophic failure
+        // The matrix is singular so return -1 to regularaize!
+        return -1;
+      } else { // Numeric was succesfull so now can set up
+        factorizationSetupSucc_ = 1;
+        nlp_->log->printf(hovSummary, "Numeric factorization succesful! \n");
+        if(refact_ == "glu") {
+          this->initializeCusolverGLU();
+          this->refactorizationSetupCusolverGLU();
+        } else if(refact_ == "rf") {
+            this->initializeCusolverRf();
+            this->refactorizationSetupCusolverRf();
+        } else { // for future -
+          assert(0 && "Only glu and rf refactorizations available.\n");
+        }
       }
-
-      checkCudaErrors(cudaMemcpy(da_, kVal_, sizeof(double) * nnz_, cudaMemcpyHostToDevice));
-
-      //re-factor here
-
-      sp_status_ = cusolverSpDgluReset(handle_cusolver_,
-                                       n_,
-                                       /* A is original matrix */
-                                       nnz_,
-                                       descr_A_,
-                                       da_,
-                                       dia_,
-                                       dja_,
-                                       info_M_);
-      //end of factor
     }
-
-
-    nlp_->runStats.linsolv.tmInertiaComp.start();
-    int negEigVal = nFakeNegEigs_;
-    nlp_->runStats.linsolv.tmInertiaComp.stop();
-    return negEigVal;
+    else { // Numeric_ != nullptr
+
+      checkCudaErrors(cudaMemcpy(da_, kVal_, sizeof(double) * nnz_,
```

Comment:
codestyle? (many other occurrences)

Reply:
> I suggest I merge the csr-hpp-dev first. And I offer to do the merging/rebasing for this if you want.

Merging `csr-hpp-dev` first makes sense. My only concern is that you still need to use fake inertia in `csr-hpp-dev` branch and I don't see the reason why is that so. This is the only potential issue I see right now.

Reply:
> > I suggest I merge the csr-hpp-dev first. And I offer to do the merging/rebasing for this if you want.
> 
> Merging `csr-hpp-dev` first makes sense. My only concern is that you still need to use fake inertia in `csr-hpp-dev` branch and I don't see the reason why is that so. This is the only potential issue I see right now.

I will remove that hack (and others) when I merge develop into this PR.